src/patterns.jl

```diff
@@ -0,0 +1,131 @@
+"""
+    AbstractPattern
+
+Abstract supertype of all sparsity pattern representations.
+
+## Type hierarchy
+```
+AbstractPattern
+├── AbstractVectorPattern: used in GradientTracer, ConnectivityTracer
+│   └── IndexSetVectorPattern
+└── AbstractHessianPattern: used in HessianTracer
+    └── IndexSetHessianPattern
+```
+"""
+abstract type AbstractPattern end
+
+#==========================#
+# Utilities on AbstractSet #
+#==========================#
+
+myempty(::Type{S}) where {S<:AbstractSet} = S()
+seed(::Type{S}, i::Integer) where {S<:AbstractSet} = S(i)
+
+""""
+    product(a::S{T}, b::S{T})::S{Tuple{T,T}}
+
+Inner product of set-like inputs `a` and `b`.
+"""
+product(a::AbstractSet{I}, b::AbstractSet{I}) where {I<:Integer} =
+    Set((i, j) for i in a, j in b)
+
+function union_product!(
+    hessian::SH, gradient_x::SG, gradient_y::SG
+) where {I<:Integer,SG<:AbstractSet{I},SH<:AbstractSet{Tuple{I,I}}}
+    hxy = product(gradient_x, gradient_y)
+    return union!(hessian, hxy)
+end
+
+#=======================#
+# AbstractVectorPattern #
+#=======================#
+
+# For use with ConnectivityTracer and GradientTracer.
+
+"""
+    AbstractVectorPattern <: AbstractPattern
+
+Abstract supertype of sparsity patterns representing a vector.
+For use with [`ConnectivityTracer`](@ref) and [`GradientTracer`](@ref).
+
+## Expected interface
+
+* `myempty(::Type{MyPattern})`: return a pattern representing a new number (usually an empty pattern)
+* `seed(::Type{MyPattern}, i::Integer)`: return an pattern that only contains the given index `i`
+* `inputs(p::MyPattern)`: return non-zero indices `i` for use with `ConnectivityTracer`
+* `gradient(p::MyPattern)`: return non-zero indices `i` for use with `GradientTracer`
+
+Note that besides their names, the last two functions are usually identical.
+"""
+abstract type AbstractVectorPattern <: AbstractPattern end
+
+"""
+$(TYPEDEF)
+
+Vector sparsity pattern represented by an `AbstractSet` of indices ``{i}`` of non-zero values.
+
+## Fields
+$(TYPEDFIELDS)
+"""
+struct IndexSetVectorPattern{I<:Integer,S<:AbstractSet{I}} <: AbstractVectorPattern
```

Comment:
I think we can use this PR to remove `AbstractSet` subtyping (because again, our custom types don't satisfy every assumption) and use duck-typing for that stuff

Reply:
My idea is that stuff like `DuplicateVector` gets to be a `DuplicatePattern` in the future.

Reply:
Okay that's fair. But then we should temporarily remove it from the API if we're gonna change its place in the hierarchy

Reply:
You mean before tagging a release?

Reply:
Yeah, let's not tag a release where DuplicateVector is an abstract set and then change it right away to a pattern